api/src/main/java/org/openmrs/api/context/Context.java

```diff
@@ -637,6 +627,38 @@ public static boolean isAuthenticated() {
 			return getAuthenticatedUser() != null;
 		}
 	}
+	
+	/**
+	 * OpenMRS provides its default authentication scheme that authenticates via DAO with OpenMRS usernames and passwords.
+	 * Any module can provide an authentication scheme override by Spring wiring a custom implementation of {@link AuthenticationScheme}.
+	 * This method would return Core's default authentication scheme unless a Spring override is provided somewhere else.
+	 * 
+	 * @return The enforced authentication scheme.
+	 */
+	public static AuthenticationScheme getAuthenticationScheme() {
+		
+		AuthenticationScheme scheme = new UsernamePasswordAuthenticationScheme();
+		
+		try {
+			scheme = Context.getServiceContext().getApplicationContext().getBean(AuthenticationScheme.class);	// manual autowiring from a module
+			log.info("An authentication scheme override was provided. Using this one in place of the OpenMRS default authentication scheme.");
+		}
+		catch(NoUniqueBeanDefinitionException e) {
+			log.error("Multiple authentication schemes overrides are being provided, this is currently not supported. Sticking to OpenMRS default authentication scheme.");
+		}
+		catch(NoSuchBeanDefinitionException e) {
+			log.debug("No authentication scheme override was provided. Sticking to OpenMRS default authentication scheme.");
+		}
+		catch(BeansException e){
+			log.error("Fatal error encountered when injecting the authentication scheme override. Sticking to OpenMRS default authentication scheme.");
+		}
+		
+		if (scheme instanceof DaoAuthenticationScheme) {
```

Comment:
Why would a get something method have a side effect of setting something?

Reply:
@dkayiwa this method is not really a getter (it doesn't return a member of `Context`). It's more akin of a factory, it's basically a wrapper around a Spring injection.
We could do this if you find it less awkward:

```java
public static AuthenticationScheme getAuthenticationScheme() {

  AuthenticationScheme scheme = new UsernamePasswordAuthenticationScheme(getContextDAO());
  
  try {
    scheme = Context.getServiceContext().getApplicationContext().getBean(AuthenticationScheme.class); // manual autowiring from a module
    if (scheme instanceof DaoAuthenticationScheme) {
      ((DaoAuthenticationScheme) scheme).setContextDao(getContextDAO());
    }
    log.info("An authentication scheme override was provided. Using this one in place of the OpenMRS default authentication scheme.");
  }
  catch(NoUniqueBeanDefinitionException e) {
    log.error("Multiple authentication schemes overrides are being provided, this is currently not supported. Sticking to OpenMRS default authentication scheme.");
  }
  catch(NoSuchBeanDefinitionException e) {
    log.debug("No authentication scheme override was provided. Sticking to OpenMRS default authentication scheme.");
  }
  catch(BeansException e){
    log.error("Fatal error encountered when injecting the authentication scheme override. Sticking to OpenMRS default authentication scheme.");
  }
  
  return scheme;
}
```

Reply:
For even a factory method that gets me an instance of an object, i would expect it to create or return an existing instance but not do anything like setting something else. This side effect does not look intuitive to me. Unless some one else disagrees. Spring injection would be a via a set, not get. 😊

Reply:
Thanks @dkayiwa. So I made an attempt at splitting clearly between the auth. scheme's getter and setter. Setting the auth. scheme in the first place is something that must be done as part of `Context`'s initialisation process and this is why it is now used as its `init-method` (see application context XML).

However it is tricky to achieve to run something like that:
```java
@Autowired(required=false)
static private void setAuthenticationScheme(AuthenticationScheme moduleAuthScheme) {

  authenticationScheme = moduleAuthScheme; 
  if (authenticationScheme == null) {
    authenticationScheme = new UsernamePasswordAuthenticationScheme();
  }
  
  if (authenticationScheme instanceof DaoAuthenticationScheme) {
    ((DaoAuthenticationScheme) authenticationScheme).setContextDao(getContextDAO());
  }
}
```
While that would be more Spring-like, it doesn't quite allow to fully handle all possible exceptions as done until now. But as I said it is tricky to make that work on static methods or on private methods.
Note also that I made sure that `setAuthenticationScheme` is private to prevent any kind of override through ever consuming the API.

Finally I tested all this at runtime again and it works fine with and without a module that Spring-wires a custom auth. scheme.